astropy/units/quantity_helper/function_helpers.py

```diff
@@ -702,8 +722,74 @@ def _check_range(range, unit):
     return range
 
 
-@function_helper
-def histogram(a, bins=10, range=None, weights=None, density=None):
+@function_helper(helps=np.histogram_bin_edges)
+def histogram_bin_edges(a, bins=10, range=None, weights=None):
+    # weights is currently unused
+    a = _as_quantity(a)
+    if not isinstance(bins, str):
+        bins = _check_bins(bins, a.unit)
+
+    if range is not None:
+        range = _check_range(range, a.unit)
+
+    return (a.value, bins, range, weights), {}, a.unit, None
+
+
+if NUMPY_LT_1_24:
```

Comment:
I think what I would do is define the numpy >= 2.0 functions above, with the regular names, and then here do
```
# For numy < 2.0, swap density and weight arguments. 
FUNCTION_HELPERS[np.histogram] = lambda a, bins=10, range=None, normed=None, weights=None, density=None: histogram(a, bins=bins, range=range, normed=normed, weights=weights, density=density)
FUNCTION_HELPERS[...
```
The main advantage is that when our minimum numpy version becomes 2.0, we can just remove the block without thinking about it. But also nice to just have one all-good implementation.

Actually, that does not quite work, since the `normed` argument was dropped. How about
```
if NUMPY_LT_2_0:
    @function_helper(helps=np.histogram)
    def histogram_pre_2_0(a, bins=10, range=None, normed=None, weights=None, density=None):
        args, kwargs, unit, out = histogram(
            a, bins=bins, range=range, weights=weights, density=density or normed
        )
        kwargs['normed'] = normed
        kwargs['density'] = density
        return args, kwargs, unit, out

...
```
Note that for better or for worse, one can override an assignment...

Reply:
> I think what I would do is define the numpy >= 2.0 functions above, 

Note that the change isn't in numpy 2.0 VS 1.x, it happened in 1.24.

> Note that for better or for worse, one can override an assignment...

Right, but I find it's maybe less confusing for readers if we don't rely on this and instead keep a clean `if/else` structure ?

Your suggested change is fine but it would not conform to [numpy's documentation](https://numpy.org/doc/1.23/reference/generated/numpy.histogram.html)
> This is equivalent to the density argument, but produces incorrect results for unequal bin widths. It should not be used.

if we pass `normed` argument internally as `density` then this behaviour will not be visible to users, which may be confusing (even though we'd be hiding a **buggy** behaviour) ? I think this would be fine, but it'd warrant a comment, at least. What do you think we should do ?

Reply:
No, my suggestion just passes it on to the *helper*, where it deals with the unit, and then changes `kwargs` so that `__array_function__` will call with the right arguments.

Note that since we did not support `normed` before, I'm also quite happy to just omit it here! It has been deprecated in numpy forever and clearly nobody in astropy complained about. How about just doing the argument swap only? (And special-case it in in the tests...)

Reply:
> No, my suggestion just passes it on to the helper, where it deals with the unit, and then changes kwargs so that __array_function__ will call with the right arguments.

Ah ! I see, then I believe that's the way to go. Doing it now. Thank you !

Reply:
My idea was to put the new definitions above (without `_impl`) and then use those here. I still think that is a bit nicer - we'll get rid of numpy 1.23 soon enough...

Reply:
gotcha, hopefully it's all good now !

Reply:
Not quite there yet: just move the actual `histogram`, etc., to before this `if NUMPY_LT_1_24` and include `@function_helper` in front of them, and then here you just override for older numpy.

Reply:
p.s. So you can remove the `else` clause!

Reply:
Ok _now_ I think it's good !